indexer/processors/bridge/l1_bridge_processor.go

```diff
@@ -0,0 +1,249 @@
+package bridge
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/indexer/database"
+	contracts "github.com/ethereum-optimism/optimism/indexer/processors/contracts_bedrock"
+	"github.com/ethereum-optimism/optimism/op-e2e/config"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// L1ProcessInitiatedBridgeEvents will query the database for new bridge events that have been initiated between
+// the specified block range. This covers every part of the multi-layered stack:
+//  1. OptimismPortal
+//  2. L1CrossDomainMessenger
+//  3. L1StandardBridge
+func L1ProcessInitiatedBridgeEvents(log log.Logger, db *database.DB, fromHeight *big.Int, toHeight *big.Int) error {
+	// (1) OptimismPortal
+	optimismPortalTxDeposits, err := contracts.OptimismPortalTransactionDepositEvents(config.L1Deployments.OptimismPortalProxy, db, fromHeight, toHeight)
+	if err != nil {
+		return err
+	}
+
+	ethDeposits := []*database.L1BridgeDeposit{}
+	portalDeposits := make(map[logKey]*contracts.OptimismPortalTransactionDepositEvent, len(optimismPortalTxDeposits))
+	transactionDeposits := make([]*database.L1TransactionDeposit, len(optimismPortalTxDeposits))
+	for i := range optimismPortalTxDeposits {
+		depositTx := optimismPortalTxDeposits[i]
+		portalDeposits[logKey{depositTx.Event.BlockHash, depositTx.Event.LogIndex}] = &depositTx
+		transactionDeposits[i] = &database.L1TransactionDeposit{
+			SourceHash:           depositTx.DepositTx.SourceHash,
+			L2TransactionHash:    types.NewTx(depositTx.DepositTx).Hash(),
+			InitiatedL1EventGUID: depositTx.Event.GUID,
+			GasLimit:             depositTx.GasLimit,
+			Tx:                   depositTx.Tx,
+		}
+
+		// catch ETH transfers to the portal contract.
+		if len(depositTx.DepositTx.Data) == 0 && depositTx.DepositTx.Value.BitLen() > 0 {
+			ethDeposits = append(ethDeposits, &database.L1BridgeDeposit{
+				TransactionSourceHash: depositTx.DepositTx.SourceHash,
+				BridgeTransfer:        database.BridgeTransfer{Tx: transactionDeposits[i].Tx, TokenPair: database.ETHTokenPair},
+			})
+		}
+	}
+
+	if len(transactionDeposits) > 0 {
+		log.Info("detected transaction deposits", "size", len(transactionDeposits))
+		if err := db.BridgeTransactions.StoreL1TransactionDeposits(transactionDeposits); err != nil {
+			return err
+		}
+		if len(ethDeposits) > 0 {
+			log.Info("detected portal ETH transfers", "size", len(ethDeposits))
+			if err := db.BridgeTransfers.StoreL1BridgeDeposits(ethDeposits); err != nil {
+				return err
+			}
+		}
+	}
+
+	// (2) L1CrossDomainMessenger
```

Comment:
Not easy to tell at a glance if these are 3 isolated steps or if there is some cross step dependencies like OptimismPortal is a dependency to L1CrossDomainmessenger being able to process

Reply:
I can add some more to the comment above but they are dependencies such that the bridge transfers can refer down to the messenger and the messenger can refer to the portal, etc, etc

Hence code blocks like this:
```go
	if len(crossDomainSentMessages) > len(transactionDeposits) {
		return fmt.Errorf("missing transaction deposit for each cross-domain message. deposits: %d, messages: %d", len(transactionDeposits), len(crossDomainSentMessages))
	}
```